docs/motivation-and-use-cases.md

```diff
@@ -0,0 +1,138 @@
+
+
+## Pinning Rings (first reported here https://github.com/ipfs/ipfs-cluster/issues/7)
+
+Actors:
+	users in an existing pinning ring
+	users interested in joining / leaving the pinning ring
+	a ring administrator (potentially multiple, potentially 0)
+
+Description: A collection of users interested in collaboratively storing a collection of data together make up an ipfs-cluster pinning ring by storing data on ipfs and adding each other as cluster peers.  When existing users want to add content, they simply add it to their ipfs nodes and pin with an appropriate replication strategy depending on the pinning ring's setup and the content being pinned.  Pinning rings that are short on storage may choose not to replicate, pinning rings looking for robustness may use high replication factors or fault tolerance strategies (see PR #268).
+
+The barrier for entry/exit is low, any user with an ipfs node and free space is welcome to join pinning rings.  When a new user wishes to enter, they provide a peer ID and multiaddr to either a ring administrator (or perhaps any participant) add the new peer to the cluster.  It is also possible that the pinning ring will allow the peer to add itself to the cluster.  The new peer then autodiscovers the rest of the ring's peers, joins the consensus and now this node provides storage to the cluster.  Similarly peers can leave the pinning ring without much warning.  Perhaps within cluster or perhaps out of band peers about to disconnect could signal this so that their pins are not totally dropped from the cluster.  Lots of storage and aggressive replication could be used to remove the fallout of sudden disconnections.
+
+While peers within a pinning ring are allowed the priveleges of a cluster peer such as the power to add and remove peers, unpin content across the cluster, and change replication factors, any cluster peer can audit the operations taking place by looking at the log maintained by consensus.  In the event a peer consistently misbehaves pinning rings may potentially wish to take action by somehow restricting this peer's access.  On a somewhat similar note pinning rings may also wish to grant "observe only" access to the cluster.
+
+Implied ipfs-cluster requirements:
+- support for dynamic cluster membership (exists today but potentially some bugs with lots of churn)
+- some kind of trust modeling support, potentially including associating permissions to operations and assigning permissions to peers.  Could make use of the proposed [capabilities service](https://github.com/ipfs/notes/issues/274).
+- support for byzantine consensus protocols sounds relevant
+- support for updating many uncoordinated nodes
```

Comment:
The main problem here is the trust model. I can think of several ways of approaching this usecase but I always find problems with how to prevent random users from altering the whole cluster. I think with Raft this is limited to who controls the cluster leader node, but Raft doesn't scale for this.

* Can think of adding authorization to RPC calls
* With the above, can think of only allowing trusted nodes to become leaders (Raft supports this)
* Can think of implementing this as an application wrapping cluster/ipfs, in which the user does not run the cluster peer but only the ipfs-daemon, and the administrator runs an associated cluster peer but under his control. This would scale with composite clusters.
* Can think of replacing the consensus layer with pubsub for this use-case, and only obeying updates signed with certain keys (perhaps for a pinning ring, the fact that the state is fully consistent among all peers rather quickly is not super important) (the more I think about this the nicer it sounds).

In any case, it is always the trust model that worries me.

Reply:
If I'm a bit bored one day I'm going to write an RFC about replacing Raft with Pubsub

Reply:
I will be super interested to read that RFC